core/src/main/java/org/elasticsearch/index/mapper/core/KeywordFieldMapper.java

```diff
@@ -0,0 +1,308 @@
+            FIELD_TYPE.freeze();
+        }
+
+        // NOTE, when adding defaults here, make sure you add them in the builder
+        public static final String NULL_VALUE = null;
+
+        public static final int IGNORE_ABOVE = -1;
+    }
+
+    public static final class KeywordFieldType extends MappedFieldType {
+
+        public KeywordFieldType() {}
+
+        protected KeywordFieldType(KeywordFieldType ref) {
+            super(ref);
+        }
+
+        public KeywordFieldType clone() {
+            return new KeywordFieldType(this);
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public String value(Object value) {
+            if (value == null) {
+                return null;
+            }
+            return value.toString();
+        }
+
+        @Override
+        public Query nullValueQuery() {
+            if (nullValue() == null) {
+                return null;
+            }
+            return termQuery(nullValue(), null);
+        }
+    }
+
+    public static class Builder extends FieldMapper.Builder<Builder, KeywordFieldMapper> {
+
+        protected String nullValue = Defaults.NULL_VALUE;
+
+        protected int ignoreAbove = Defaults.IGNORE_ABOVE;
+
+        public Builder(String name) {
+            super(name, Defaults.FIELD_TYPE);
+            builder = this;
+        }
+
+        @Override
+        public Builder searchAnalyzer(NamedAnalyzer searchAnalyzer) {
+            super.searchAnalyzer(searchAnalyzer);
+            return this;
+        }
+
+        public Builder ignoreAbove(int ignoreAbove) {
+            this.ignoreAbove = ignoreAbove;
+            return this;
+        }
+
+        @Override
+        public KeywordFieldMapper build(BuilderContext context) {
+            // if the field is not analyzed, then by default, we should omit norms and have docs only
+            // index options, as probably what the user really wants
+            // if they are set explicitly, we will use those values
+            // we also change the values on the default field type so that toXContent emits what
+            // differs from the defaults
+            if (fieldType.indexOptions() != IndexOptions.NONE) {
+                defaultFieldType.setOmitNorms(true);
+                defaultFieldType.setIndexOptions(IndexOptions.DOCS);
+                if (!omitNormsSet && fieldType.boost() == 1.0f) {
+                    fieldType.setOmitNorms(true);
+                }
+                if (!indexOptionsSet) {
+                    fieldType.setIndexOptions(IndexOptions.DOCS);
+                }
+            }
+            setupFieldType(context);
+            if (context.indexCreatedVersion().onOrAfter(Version.V_3_0_0)
+                    && fieldType.indexOptions().compareTo(IndexOptions.DOCS_AND_FREQS_AND_POSITIONS) >= 0) {
+                throw new MapperParsingException("Field [" + name
```

Comment:
actually this would allow someone to bypass the exception by creating the mapper directly by using the builder (some plugins do that). So I moved it, but to the KeywordFieldMapper constructor to make sure that nothing can by-pass this check. This is already how analyzed string fields ensure that doc values have not been enabled.